### DISCUSSION TITLE
Nested many to many relationship

### DISCUSSION THREAD

**Discussion by erasebegin** [prisma/prisma#16875] [Q&A]
I have been working with a setup that I *thought* was working, but it turns out it was only partially working. I am aiming to have a table called `Person` with a column called `people` which is a `Person[ ]`. The idea is that a person can be related to many people, so I had it set up like this:

```
model Person {
  id               String       @id @default(uuid())
  firstName        String?
  lastName         String?
  Person           Person?      @relation("RelatedPeople", fields: [personId], references: [id])
  personId         String?
  people           Person[]     @relation("RelatedPeople")
}
``` 

But I only realised recently that this would mean a person could only refer to another person once. If another person tried to refer to that same person, it would be removed from the first person and attached to the second person. So what I actually need is a many-many relationship where many people can be related to many people.

I was able to achieve this recently with another table, but that was not nested many-many so Prisma's formatter was able to help me set up the correct relationship. When I try to use the formatter here however, it just gives me the above setup. I understand that for many-many relationships a join table needs to be created under the hood, and I can see that that's what Prisma has been doing for the other relationships, but I don't know how to do it in a nested/recursive way.

Please could you data wizards please help me to resolve this? 🙏🙏🙏

**Comment by pantharshit00** [comment]
Hello @erasebegin,

I think what you asking for is a kind of a follower-following model from social apps where people add each other from both ways. You can do so by using the following schema:
```prisma
model Person {
  id        String   @id @default(uuid())
  firstName String?
  lastName  String?
  followers Person[] @relation(name: "followers")
  following Person[] @relation(name: "followers")
}
```

This will create the following tables under the hood:
<img width="157" alt="image" src="https://user-images.githubusercontent.com/22195362/208386349-b77c1b41-3507-4327-a305-68f2b2078615.png">

`_followers` table is automatically generated for the relation using the `name` field which we provided. In that table it is possible to add relations from both ends and it is all managed by Prisma:
<img width="811" alt="image" src="https://user-images.githubusercontent.com/22195362/208386543-3657044b-5ae6-4b8f-8600-b23e47df127a.png">



Let me know if this was what you were looking for or your usecase is something else.

**Reply by erasebegin** [reply]
Hi, thank you so much for your response! This looks like it would work, but I'm not sure it fits my use case. In my case there doesn't need to be a distinction between the people following a person and the people that person is following. It is just that once a connection is established between two people, it works both ways. So I can look up person A and it will show that they are linked with person B, or I can look up person, B and it will show that they are linked with person A.

To put it in context, it is a system for recording details about a person. As well as contact details etc, the user will also have the option to select other people that they are connected to in some way. So for example the user is recording details for Jack and in the same form, they can select from a list of people in the database that will determine which people are connected to Jack. Maybe one is his father, another is his legal guardian etc. So everyone that is selected is a "follower" of Jack, or it could equally be said that Jack is a follower of all of them.

So could I just use the `following` column without the `followers` column? Does that still set up a join table? Apologies I find this all very confusing 😅

**Reply by pantharshit00** [reply]
Hello again!

I don't think this would be achieved by the current client API. We also currently force you to provide a back relation for every relation. 

The model I suggested would work for this case but if you will do RAW SQL query, you would be able to aggregate data from both side. Right now that is not possible with Prisma. 

I have created this feature request to track this use case: https://github.com/prisma/prisma/issues/16977

**Reply by erasebegin** [reply]
Ahh alright, I'll have to ask one of the SQL wizards at work how to achieve this then, thank you so much for your help 🙏 I hope this becomes a feature eventually as I'm sure I won't be the last person to need this! (although it seems I'm the first, so maybe not so high priority 😅)

**Reply by erasebegin** [reply]
@pantharshit00 sorry to bother you again, but I'm looking at this and I'm not sure what to do. I've gone into the Supabase database and created a join table, but when I do `prisma db pull` it thows an error and says there are no columns in the join table I've created. How do I use a join table like this with Prisma even though Prisma doesn't support this kind of self-referencing relationship? 

I can make the table in the db, but then how to I use Prisma to write to that table if I can't get it into the Prisma schema?

**Reply by erasebegin** [reply]
I've found a temporary solution which involves creating a string containing all user ids that are connected to that person and then breaking apart the string and fetching data for all ids at the time that the table is retrieved. I hope that Prisma can implement something to help with this in future 🙏

**Reply by janpio** [reply]
Do you not need meta data on these relationships anyway @erasebegin? In an implici m:n relation you can not sore the information if someone is "father" or "legal guardian" - you only get "these two are connected".

**Reply by erasebegin** [reply]
The way that relationships are handled is a bit more basic than that. Rather than saying X is Y's father, we just say X has `role` of `father`. Which is not ideal, but it takes the burden off of the user where they would have to define these relationships in every direction.

**Reply by janpio** [reply]
How do you represent that with an implicit m:n then?

Would this be covered by https://github.com/prisma/prisma/issues/14370?

**Reply by erasebegin** [reply]
Yes that seems related. I'm not sure what is meant by implicit, but I have achieved the desired result with two different tables where I say:

```
model Case {
    people  Person[ ]
}

model Person {
    cases  Case[ ]
}
```

I just want to set up the same sort of thing, but in a self-relating way. The way I've managed to achieve it is now mapping through ids and making many calls to the database which is far from ideal (but seems to work)

**Reply by janpio** [reply]
"implicit m:n" in this case describes an m:n relation that does not have an explicit join table in your Prisma schema: https://www.prisma.io/docs/concepts/components/prisma-schema/relations/many-to-many-relations#implicit-many-to-many-relations

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
